woocommerce/woocommerce-android
# Add Debounced Scrolling to Analytics Usage Tracking

Closes #5756. Continuation of https://github.com/woocommerce/woocommerce-android/pull/5964. Please review that PR first. Thank you!

⚠️ I'm keeping this as a draft since I just wanted to ask for inputs first. This still needs some touch-ups. And this is my first time using Kotlin Flow so I could use some suggestions. 🙂 

## Background

In #5964, I mentioned that I have not included scrolling as one of the UI events that will call [`MyStoreStatsUsageTracksEventEmitter.interacted()`](https://github.com/woocommerce/woocommerce-android/blob/217f94bd9ca690de3f103f8de9a63f197368b118/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/mystore/MyStoreStatsUsageTracksEventEmitter.kt#L60). The reason is:

> I have not added scrolling yet. I wanted to call `interact()` only when we detect that the user **started** scrolling. But it doesn't look like [`NestedScrollView` ](https://developer.android.com/reference/androidx/core/widget/NestedScrollView) has an event that provides this. It only has [`OnScrollChangeListener`](https://developer.android.com/reference/androidx/core/widget/NestedScrollView.OnScrollChangeListener) which is triggered way too many times to be useable. 

Here's a video showing how many events are being triggered:


https://user-images.githubusercontent.com/198826/156255669-3ee0ef1f-224d-450d-bd9d-1b8a77a78954.mov



## Solution

I wanted this to be similar to iOS where we only call `interacted()` when the user started scrolling. I researched for options but it looked like this was going to get really complex really fast. 😬 There are suggestions like subclassing `NestedScrollView`. But even then, I'm not even sure if that's worth it. 

A `throttleFirst()` operator (https://github.com/Kotlin/kotlinx.coroutines/issues/1446) seems to be the most appropriate solution. But it doesn't look like it's available natively yet. There are suggestions about how to implement it but I don't know if we want to go to all that trouble. 

In the end, I decided to use `debounce`. It's not the best but the resulting code seems simple enough. And with a debounce duration of 1000, the effect feels similar. 

## Testing

Same steps as #5964 but now we can include scrolling.

1. Log in
2. Navigate to My Store
3. Do a few of the following:
     - Scroll 🆕 
     - Pull-to-refresh
     - Tapping on the bars in the chart
     - Changing the tab
     - Navigating to the My Store tab
     - Tapping on a product in the Top Performers list
5. Confirm that after performing at least 5 of the interactions within more than 10 seconds, a `used_analytics` Tracks event is logged.

## Reviewing

Only 1 reviewer is needed but anyone can review.

### Questions

1. How is the implementation? Is there a simpler way of doing this?
2. Any concerns with how I'm using Flow here? I was concerned about not correctly _cancelling_ the subscription [here](https://github.com/woocommerce/woocommerce-android/blob/625ad89c13da9f060a2d9d2f5f16f0f33bf4d279/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/mystore/MyStoreFragment.kt#L142-L145) but it looks like it works fine?
3. Will this affect scrolling performance? I was curious whether this should be executed in `.io` but I currently don't know how to do that with Flow. 🤪 

## Author Checklist
